src/main/java/org/folio/circulation/services/AllowedServicePointsService.java

```diff
@@ -32,7 +60,92 @@ private CompletableFuture<Result<Map<RequestType, Set<String>>>> getAllowedServi
+    RequestPolicy requestPolicy) {
+
+    log.debug("extractAllowedServicePoints:: parameters: requestPolicy={}", requestPolicy);
+
+    var allowedServicePointsInPolicy = requestPolicy.getAllowedServicePoints();
+    if (allowedServicePointsInPolicy.isEmpty()) {
+      log.info("extractAllowedServicePoints:: allowedServicePointsInPolicy is empty");
+
+      return fetchAllowedServicePoints(requestPolicy);
+    }
+
+    Set<String> allowedServicePointsIdsSet = allowedServicePointsInPolicy.values().stream()
+      .flatMap(Collection::stream)
+      .map(UUID::toString)
+      .collect(Collectors.toSet());
+
+    return servicePointRepository.filterIdsByServicePointsAndPickupLocationExistence(
+      allowedServicePointsIdsSet)
+        .thenApply(r -> r.next(filteredSpIds -> includeOnlyExistingServicePoints(
+          allowedServicePointsInPolicy, filteredSpIds, requestPolicy)));
+  }
+
+  private CompletableFuture<Result<Map<RequestType, Set<String>>>> fetchAllowedServicePoints(
+    RequestPolicy requestPolicy) {
+
+    log.debug("fetchAllowedServicePoints:: parameters: requestPolicy={}", requestPolicy);
+
+    List<RequestType> allowedTypes = Arrays.stream(RequestType.values())
+      .filter(requestPolicy::allowsType)
+      .toList();
+    if (allowedTypes.isEmpty()) {
+      log.info("fetchAllowedServicePoints:: allowedTypes is empty");
+      return ofAsync(new EnumMap<>(RequestType.class));
+    }
+    log.info("fetchAllowedServicePoints:: allowedTypes={}", allowedTypes);
+    return servicePointRepository.fetchPickupLocationServicePointsIds()
+      .thenCompose(r -> r.after(servicePointIds -> {
+        Map<RequestType, Set<String>> allowedServicePoints = new EnumMap<>(RequestType.class);
+        allowedTypes.forEach(requestType -> allowedServicePoints.put(requestType, servicePointIds));
+        log.info("fetchAllowedServicePoints:: result: {}", () -> asJson(allowedServicePoints));
+
+        return ofAsync(allowedServicePoints);
+      }));
+
+  }
+
+  private Result<Map<RequestType, Set<String>>> includeOnlyExistingServicePoints(
+    Map<RequestType, Set<UUID>> allowedServicePointsInPolicy, Set<String> relevantSpIds,
+    RequestPolicy requestPolicy) {
+
+    log.debug("includeOnlyExistingServicePoints:: parameters: allowedServicePointsInPolicy={}, " +
+      "relevantSpIds={}, requestPolicy={}", () -> asJson(allowedServicePointsInPolicy),
+      () -> asJson(relevantSpIds), () -> asJson(requestPolicy));
+
+    final Map<RequestType, Set<String>> allowedServicePoints = new EnumMap<>(RequestType.class);
+    allowedServicePointsInPolicy.forEach((requestType, uuidSet) -> {
+      if (requestPolicy.allowsType(requestType)) {
+        allowedServicePoints.put(requestType,
+          uuidSet.stream()
+            .map(UUID::toString)
+            .filter(relevantSpIds::contains)
+            .collect(Collectors.toSet()));
+      }
+    });
+
+    return succeeded(allowedServicePoints);
```

Comment:
No need to wrap return value into a Result

Reply:
done